Samples/V1.0Samples/LocalMediaSamples/AudioVideoPlaybackBot/FrontEnd/AVPFrontEnd.csproj

```diff
@@ -1,75 +1,77 @@
-    <Reference Include="Microsoft.Skype.Internal.Media.H264, Version=0.0.0.0, Culture=neutral, processorArchitecture=AMD64">
-      <SpecificVersion>False</SpecificVersion>
-      <HintPath>..\..\..\..\packages\Microsoft.Skype.Bots.Media\1.17.0.39-alpha\src\skype_media_lib\Microsoft.Skype.Internal.Media.H264.dll</HintPath>
-    </Reference>
-  </ItemGroup>
-  <ItemGroup>
-    <PackageReference Include="Microsoft.AspNet.WebApi.Owin" Version="5.2.7" />
-    <PackageReference Include="Microsoft.Graph.Communications.Calls.Media" Version="1.2.0.850" />
-    <PackageReference Include="Microsoft.Owin.Hosting" Version="4.0.1" />
-    <PackageReference Include="Microsoft.Skype.Bots.Media" Version="1.17.0.39-alpha" />
-    <Reference Include="System.Data.Entity" />
-  </ItemGroup>
-  <ItemGroup>
-    <ProjectReference Include="..\..\..\..\Common\Sample.Common\Sample.Common.csproj">
-      <Project>{3268e59c-90dc-4d7b-97ea-a1dbb2716df3}</Project>
-      <Name>Sample.Common</Name>
-    </ProjectReference>
-  </ItemGroup>
-  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+  <Import Project="$(MSBuildThisFileDirectory)..\..\..\..\Graph.props" />
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProjectGuid>{7C0DE005-FD92-4AEE-A6A2-03954F9DA631}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>Sample.AudioVideoPlaybackBot.FrontEnd</RootNamespace>
+    <AssemblyName>Sample.AudioVideoPlaybackBot.FrontEnd</AssemblyName>
+    <TargetFrameworkVersion>v4.6.2</TargetFrameworkVersion>
+    <FileAlignment>512</FileAlignment>
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'">
+    <DebugSymbols>true</DebugSymbols>
+    <OutputPath>bin\x64\Debug\</OutputPath>
+    <DefineConstants>TRACE;DEBUG;PLAY_MEDIA_FILE</DefineConstants>
+    <DebugType>full</DebugType>
+    <PlatformTarget>x64</PlatformTarget>
+    <LangVersion>latest</LangVersion>
+    <ErrorReport>prompt</ErrorReport>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x64'">
+    <OutputPath>bin\x64\Release\</OutputPath>
+    <DefineConstants>PLAY_MEDIA_FILE;TRACE</DefineConstants>
+    <Optimize>true</Optimize>
+    <DebugType>pdbonly</DebugType>
+    <PlatformTarget>x64</PlatformTarget>
+    <ErrorReport>prompt</ErrorReport>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>

```

Comment:
Yeah, it's used in BotMediaStream.cs:

                unsafe
                {
                    Buffer.MemoryCopy(buffer.Data.ToPointer(), clonedDataPointer.ToPointer(), length, length);
                }

Reply:
Is that ok @zihzhan ?

Reply:
Not quiet understand the CloneVideoBuffer usage here, can you please elaborate a little bit more?

Reply:
In the spotlighting scenario, we forward a meeting attendee's incoming video to the outgoing video socket of the bot.  But we can't simply pass a pointer to the incoming video frame memory buffer, since that memory will be reused by the next incoming video frame.  So we have to make a copy (clone) of the video frame memory buffer to pass to the outgoing video socket.